File: python/Isis/CutManager.py
#///////////////////////////////////////////////////////////
#//                                                       //
#//  Python modules                                       //
#//                                                       //
#// ----------------------------------------------------- //
#//                                                       //
#//  AUTHOR: Miguel Ramos Pernas                          //
#//  e-mail: miguel.ramos.pernas@cern.ch                  //
#//                                                       //
#//  Last update: 04/04/2016                              //
#//                                                       //
#// ----------------------------------------------------- //
#//                                                       //
#//  Description:                                         //
#//                                                       //
#//  Class to storage cuts present in a cuts-file. This   //
#//  files have to be written in the way:                 //
#//                                                       //
#//  cut1 = j > 0 and b > 1                               //
#//                                                       //
#//  cut2 = $cut1$ or c == 0                              //
#//                                                       //
#//  The symbol < $ > before and after a word means that  //
#//  is a cut, so the class is going to get that and      //
#//  place it instead. This class is going to search      //
#//  long as no key has the same string value as it,      //
#//  being situated at the beginning of the line.         //
#//                                                       //
#// ----------------------------------------------------- //
#///////////////////////////////////////////////////////////


from collections import OrderedDict


#_______________________________________________________________________________

class CutManager:

    def __init__( self, file_name = False ):
        ''' Constructor given the name of the cuts file '''
        self.Options = { 'and': '&&', 'or': '||' }
        self.CutFile = []
        self.CutList = OrderedDict()
        if file_name:
            self.Open( file_name )

    def __getitem__( self, key ):
        ''' Gets the cut related to the key given '''
        return self.CutList[ key ]

    def Clear( self ):
        ''' Clears the content obtained from the file as well as the cuts booked '''
        self.CutList = OrderedDict()
        self.CutFile = []
        
    def ClearFile( self ):
        ''' Clears the content obtained from the file '''
        self.CutFile = []

    def BookCut( self, name, prt = True ):
        ''' Books a new cut and returns it '''
        if name in self.CutList:
            print 'Cut with name <', name, '> already booked'
            return False
        cut = self.GetCut( name )
        if cut:
            self.CutList[ name ] = cut
            if prt:
                print 'Booked new cut <', name, '>:', self.CutList[ name ]
        return cut

    def GetCut( self, name ):
        ''' Gets the Cut named < name > in the file attached to this class '''
        cuts = False
        for line in self.CutFile:
            first, second = line.split( '=' )
            while first.startswith( ' ' ):  first  = first[ 1: ]
            while first.endswith( ' ' ):    first  = first[ :-1 ]
            while second.startswith( ' ' ): second = second[ 1: ]
            while second.endswith( ' ' ):   second = second[ :-1 ]
            if first == name:
                cuts = second
                while cuts[ 0 ] == ' ': cuts = cuts[ 1: ]
                ifirst = 0
                while '$' in cuts:
                    ifirst = cuts.find( '$', ifirst )
                    ilast  = cuts.find( '$', ifirst + 1 )
                    newcut = cuts[ ifirst : ilast + 1 ]
                    cuts   = cuts.replace( newcut, self.GetCut( newcut[ 1:-1 ] ) )
                for el in self.Options:
                    while el in cuts:
                        cuts = cuts.replace( el, self.Options[ el ] )
                    break
        if cuts:
            return '(' + cuts + ')'
        else:
            print 'WARNING: cut with name <', name, '> does not exist.'
            return False

    def Open( self, file_name ):
        ''' Opens a new file and gets its content. In this method the lines are checked to see
        if they have the correct format. '''
        self.CutFile = []
        ifile = open(file_name)
        for il, line in enumerate( ifile ):
            if line[ -1 ] == '\n':
                line = line[ :-1 ]
            if line != '' and line[ 0 ] != '#':
                pos = line.find( '=' )
                if line.find( '=', pos + 1 ) != -1:
                    print 'ERROR: Found two assignment operators in line <', il, '>'
                    return
                if pos != -1:
                    ''' Checks that the cut name is free of whitespaces '''
                    pos = line[ :pos ]
                    while pos.startswith( ' ' ):
                        pos = pos[ 1: ]
                    while pos.endswith( ' ' ):
                        pos = pos[ :-1 ]
                    if pos.find( ' ' ) != -1:
                        print 'ERROR: The cut defined in line <', il + 1, '> has whitespaces on its name'
                        self.CutFile = []
                        return
                    ''' Checks the matching of the < $ > symbols in the expression '''
                    pos, newpos = 0, 0
                    while '$' in line[ newpos: ]:
                        pos     = line.find( '$', newpos )
                        pos    += 1
                        newpos  = line.find( '$', pos )
                        sbl     = line[ pos:newpos ]
                        if newpos == -1 or ' ' in sbl:
                            print 'ERROR: Mismatched < $ > symbol in line <', il + 1, '>'
                            self.CutFile = []
                            return
                        newpos += 1
                        
                else:
                    print 'WARNING: Line number <', il + 1, '> not a cut line; must be commented (starting by #)'
                self.CutFile.append( line )
        ifile.close()

    def Print( self ):
        ''' Prints the cuts booked in the class '''
        for key in self.CutList:
            print key, '->', self.CutList[ key ]

    def RemoveCut( self, kw ):
        ''' Removes the cut identified by the key < kw > '''
        del self.CutList[ key ]

File: General/source/CutManager.cpp
///////////////////////////////////////////////////////////
//                                                       //
//  General package                                      //
//                                                       //
// ----------------------------------------------------- //
//                                                       //
//  AUTHOR: Miguel Ramos Pernas                          //
//  e-mail: miguel.ramos.pernas@cern.ch                  //
//                                                       //
//  Last update: 04/04/2016                              //
//                                                       //
// ----------------------------------------------------- //
//                                                       //
//  Description:                                         //
//                                                       //
//  Class to storage cuts present in a cuts-file. This   //
//  files have to be written in the way:                 //
//                                                       //
//  cut1 = j > 0 and b > 1                               //
//                                                       //
//  cut2 = $cut1$ or c == 0                              //
//                                                       //
//  The symbol < $ > before and after a word means that  //
//  is a cut, so the class is going to get that and      //
//  place it instead. This class is going to search      //
//  for the given keys, so one can comment anything as   //
//  long as no key has the same string value as it,      //
//  being situated at the beginning of the line.         //
//                                                       //
// ----------------------------------------------------- //
///////////////////////////////////////////////////////////


#include "CutManager.h"

#include <algorithm>
#include <cstring>
#include <iomanip>
#include <limits>
#include <sstream>
#include <utility>


//_______________________________________________________________________________       


// -- CONSTRUCTORS AND DESTRUCTOR

//_______________________________________________________________________________       
// Main constructor
General::CutManager::CutManager( const std::string &file_path ) {
  fOptions[ "and" ] = "&&" ;
  fOptions[ "or"  ] = "||" ;
  if ( file_path.size() )
    this -> Open( file_path );
}

//_______________________________________________________________________________       
// Destructor
General::CutManager::~CutManager() { if ( fFile.is_open() ) fFile.close(); }

//_______________________________________________________________________________       


// -- METHODS

//_______________________________________________________________________________       
// Appends to the given vector the cuts of the current class
void General::CutManager::AppendCuts( std::vector<std::string> &vector ) {
  std::vector<std::string> appvec( fCuts.size() );
  auto itv = appvec.begin();
  auto itm = fCuts.begin();
  while ( itm != fCuts.end() )
    *itv++ = itm++ -> second;
  vector.insert( vector.end(), appvec.begin(), appvec.end() );
}

//_______________________________________________________________________________       
// Appends to the given map the cuts of the current class. If the cut already
// exists, a warning message is displayed and the cut is replaced.
void General::CutManager::AppendCuts( std::map<std::string, std::string> &map ) {
  for ( auto itm = fCuts.begin(); itm != fCuts.end(); ++itm ) {
    if ( map.find( itm -> first ) != map.end() )
      std::cout << "WARNING: Replacing existing cut < " << itm -> first << " >" << std::endl;
    map[ itm -> first ] = itm -> second;
  }
}

//_______________________________________________________________________________       
// Almacenates a new cut from the cuts-file
std::string General::CutManager::BookCut( const std::string &key, const bool &print ) {
  if ( fCuts.find( key ) != fCuts.end() ) {
    std::cout << "Cut with name < " << key << " > already booked" << std::endl;
    return "";
  }
  std::string cut = this -> GetCut( key );
  if ( cut.size() ) {
    fCuts.insert( std::make_pair( key, cut ) );
    if ( print )
      std::cout << "Booked new cut < " << key << " >: " << fCuts[ key ] << std::endl;
  }
  return cut;
}

//_______________________________________________________________________________       
// Functions that gets a cut from the cuts-file
std::string General::CutManager::GetCut( const std::string &key ) {

  // Each time a cut is obtained the pointer to the file is set to its start
  fFile.clear();
  fFile.seekg( 0 );

  std::string cuts, newcut, sstr;
  std::string::size_type pos, ifirst, ilast, fpos;
  bool cond( true ), mismatch;

  while ( cond && std::getline( fFile, cuts ) ) {

    // Checks whether the line is empty
    if ( !cuts.empty() ) {

      // If there are whitespaces before the keyword, they are omitted
      while ( cuts[ 0 ] == ' ' ) cuts.erase( 0, 1 );

      if ( !cuts.compare( 0, key.size(), key ) ) {

	cuts = cuts.substr( cuts.find( '=' ) + 1 );

	while ( cuts[ 0 ] == ' ' ) cuts.erase( 0, 1 );

	// If another cut is found it is replaced by its content. The position at the file
	// is saved and retrieved, in order to properly scan the file.
	while ( ( ifirst = cuts.find( '$' ) ) != std::string::npos ) {

	  mismatch = false;
	  ilast    = cuts.find( '$', ifirst + 1 );

	  if ( ilast == std::string::npos ) {
	    mismatch = true;
	    sstr.clear();
	  }
	  else {
	    sstr = cuts.substr( ifirst + 1, ilast - ifirst - 1 );
	    if ( sstr.find( ' ' ) != std::string::npos )
	      mismatch = true;
	  }
	  
	  if ( mismatch ) {
	    std::cerr <<
	      "ERROR: Mismatched < $ > symbol when scanning cut < "
		      << key << " >" << std::endl;
	    return std::string();
	  }

	  fpos  = fFile.tellg();

	  // The new cut is introduced between parentheses
	  newcut = "()";
	  newcut.insert( 1, std::string( this -> GetCut( sstr ) ) );
	  
	  cuts.replace( ifirst, ilast + 1 - ifirst, newcut );
	  fFile.seekg( fpos );
	}

	cond = false;
      }
    }

    if ( fFile.eof() && cond ) {
      std::cerr << "ERROR: Cut with name < " << key << " > does not exist." << std::endl;
      return 0;
    }
  }
  // Changes the name of the conditionals in the cuts set
  for ( std::map<std::string, std::string>::iterator it = fOptions.begin();
	it != fOptions.end(); it++ )
    while ( ( pos = cuts.find( it -> first ) ) != std::string::npos )
      cuts.replace( pos, ( it -> first ).size(), it -> second );

  return cuts;
}

//_______________________________________________________________________________
// Appends all the cuts in the same string using the given statement (< && > by
// default)
std::string General::CutManager::MakeMergedCut( std::string joinop ) {
  std::string gcut( '(' + fCuts.begin() -> second + ')' );
  joinop = ' ' + joinop + " (";
  auto it = fCuts.begin();
  for ( ++it; it != fCuts.end(); ++it )
    gcut.append( joinop + it -> second + ')' );
  return gcut;
}

//_______________________________________________________________________________       
// Opens the file in the given path. The old file will be closed first.
void General::CutManager::Open( const std::string &file_path ) {
  if ( fFile.is_open() )
    fFile.close();
  fFile.open( file_path.c_str() );
  std::string line, str;
  size_t nl = 0, pos, newpos;
  while ( std::getline( fFile, line ) ) {
    ++nl;
    if ( line.size() )
      if ( line.front() != '#' ) {
	if ( ( pos = line.find( '=' ) ) != std::string::npos ) {

	  // Checks whether the cut name is free of whitespaces
	  str = line.substr( 0, pos );
	  while ( str.front() == ' ' )
	    str.erase( str.begin() );
	  while ( str.back() == ' ' )
	    str.erase( str.end() - 1 );
	  if ( str.find( ' ' ) != std::string::npos ) {
	    std::cerr << "ERROR: The cut defined in line < " << nl <<
	      " > has whitespaces on its name" << std::endl;
	    return;
	  }

	  // Checks the matching of the < $ > symbols in the expression
	  while ( ( pos = line.find( '$', ++pos ) ) != std::string::npos ) {
	    newpos = line.find( '$', ++pos );
	    str    = line.substr( pos, newpos - pos );
	    if ( newpos == std::string::npos || str.find( ' ' ) != std::string::npos ) {
	      std::cerr << "ERROR: Mismatched < $ > symbol in line < " << nl << " >" << std::endl;
	      return;
	    }
	    pos = newpos + 1;
	  }
	}
	else
	  std::cout << "WARNING: Line number < " << nl <<
	    " > not a cut line; must be commented (starting by #)" << std::endl;
      }
  }
  fFile.clear();
  fFile.seekg( 0 );
}

//_______________________________________________________________________________       
// Prints the cuts attached to this class
void General::CutManager::Print() {
  size_t maxsize = std::max_element( fCuts.begin(),
				     fCuts.end(),
				     [] ( const std::pair<std::string, std::string> &p1,
					  const std::pair<std::string, std::string> &p2 ) {
				       return p1.first.size() < p2.first.size(); } ) -> first.size();
  for ( auto it = fCuts.begin(); it != fCuts.end(); it++ )
    std::cout << std::left << std::setw( maxsize ) <<
      it -> first  << " => " << it -> second << std::endl;
}
